### DISCUSSION TITLE
Execution time floats with task runtime

### DISCUSSION THREAD

**Discussion by hailwood** [spatie/laravel-cronless-schedule#11] [Ideas]
Say we start the package at 09:34:00 and it takes 2s to run all the cron commands
The next execution will start at 09:35:02 and assuming subsequent runs also take 2s
The next execution will start at 09:36:04
The next execution will start at 09:37:06

I was going to try a PR for anchor seconds so if you said --anchor 0 and started running it at 09:37:41 it wouldn't do it's first run until 09:38:00 but if the time floats anyway there's not much point in that PR.

Would you consider this a bug or is it intended?

**Comment by freekmurze** [comment]
Thanks for your suggestion. I haven't needed anchoring the time in my projects. I'm thinking the current behaviour is fine.

**Comment by robin-vendredi** [comment]
Thank you both for this package and for this comment. We ran into the same thing that was causing an issue for us. 

My understanding is that `schedule:run` checks what commands needs to be run, run them, and exit. The run-cronless package will then call `schedule:run` again one minute after **the end** of the `schedule:run` (not the beginning). Since it can take time to run the commands, this means the ticking start time slowly drifts.

To test it out, I created a command that sleeps for 10s scheduled to run every 10s. This is my output:

> [2023-08-24 17:40:31] - Running schedule:run...
>  2023-08-24 17:40:31 Running ['artisan' dev:test-cmd] ................................................ 10,424.12ms FAIL
> [2023-08-24 17:40:41] - schedule:run finished.
>
> [2023-08-24 17:41:41] - Running schedule:run...
>  2023-08-24 17:41:41 Running ['artisan' dev:test-cmd] ................................................ 10,373.91ms FAIL
> [2023-08-24 17:41:52] - schedule:run finished.
>
> [2023-08-24 17:42:52] - Running schedule:run...
>  2023-08-24 17:42:52 Running ['artisan' dev:test-cmd] ................................................ 10,384.52ms FAIL
> [2023-08-24 17:43:02] - schedule:run finished.
>
> [2023-08-24 17:44:02] - Running schedule:run...
>  2023-08-24 17:44:02 Running ['artisan' dev:test-cmd] ................................................ 10,393.55ms FAIL
> [2023-08-24 17:44:12] - schedule:run finished.

In this example, `schedule:run` hasn't run on minute `17:43`, so any command scheduled for that exact time wouldn't run.

This was causing an issue for us as we have a number of commands scheduled with `->dailyAt("10:00")` (for ex), and on some days all the commands scheduled for a given minute wouldn't run.

We fixed this by adding `->runInBackground()` to all scheduled commands, that way the `schedule:run` exits almost immediately and there is no more drift.

If my understanding is correct this is a tricky footgun to investigate - maybe that should be mentioned in the docs? I'd be happy to open a PR for that if you'd like.

Thanks for the package!
